File: 0323-number-of-connected-components-in-an-undirected-graph/0323-number-of-connected-components-in-an-undirected-graph.cpp
// TC- O(n+(e*a(n)))
// SC- O(n)

class UnionFind{
public:
    // SC- O(n)
    vector<int> parent, rank;
    int components;

    UnionFind(int n){
        // TC- O(n)
        parent.resize(n);
        for(int i=0; i<n; i++)
            parent[i]=i;
        rank.resize(n, 0);
        components = n;
    }

    int find(int x){
        // TC- O(a(n)) amortized where a is the inverse Auckermann function- grows slowly with n
        // with path compression
        if(parent[x]!=x)
            parent[x] = find(parent[x]);
        
        return parent[x];
    }

    void unite(int x, int y){
        // TC- O(a(n)) amortized where a is the inverse Auckermann function- grows slowly with n
        // union by rank
        int px = find(x);
        int py = find(y);
        
        if(px==py)
            return;
        
        if(rank[px]<rank[py])
            parent[px] = py;
        else if(rank[px]>rank[py])
            parent[py] = px;
        else{
            parent[px] = py;
            rank[py]++;
        }

        components--;
    }
};

class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        UnionFind uf(n);
        for(auto& edge:edges){
            uf.unite(edge[0], edge[1]);
        }

        return uf.components;
    }
};

File: 0323-number-of-connected-components-in-an-undirected-graph/0323-number-of-connected-components-in-an-undirected-graph.py
# TC- O(n+e)
# SC- (n+e)

class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        
        adj_mat = defaultdict(list)
        for edge in edges:
            adj_mat[edge[0]].append(edge[1])
            adj_mat[edge[1]].append(edge[0])
        
        num = 0
        visited = set()

        def dfs(i):
            visited.add(i)
            for neigh in adj_mat[i]:
                if neigh not in visited:
                    dfs(neigh)

        for i in range(n):
            if i not in visited:
                num += 1
                dfs(i)
        return num